Changes in module 'nf-core/kallisto/quant'
--- modules/nf-core/kallisto/quant/main.nf
+++ modules/nf-core/kallisto/quant/main.nf
@@ -2,7 +2,7 @@
     tag "$meta.id"
     label 'process_high'
 
-    conda "${moduleDir}/environment.yml"
+    conda "bioconda::kallisto=0.48.0"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/kallisto:0.48.0--h15996b6_2':
         'biocontainers/kallisto:0.48.0--h15996b6_2' }"

--- modules/nf-core/kallisto/quant/environment.yml
+++ /dev/null
@@ -1,7 +0,0 @@
-name: kallisto_quant
-channels:
-  - conda-forge
-  - bioconda
-  - defaults
-dependencies:
-  - bioconda::kallisto=0.48.0

************************************************************
