Changes in module 'nf-core/rseqc/innerdistance'
--- modules/nf-core/rseqc/innerdistance/main.nf
+++ modules/nf-core/rseqc/innerdistance/main.nf
@@ -2,10 +2,10 @@
     tag "$meta.id"
     label 'process_medium'
 
-    conda "bioconda::rseqc=3.0.1 'conda-forge::r-base>=3.5'"
+    conda "bioconda::rseqc=5.0.3"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/rseqc:3.0.1--py37h516909a_1' :
-        'biocontainers/rseqc:3.0.1--py37h516909a_1' }"
+        'https://depot.galaxyproject.org/singularity/rseqc:5.0.3--py39hf95cd2a_0' :
+        'biocontainers/rseqc:5.0.3--py39hf95cd2a_0' }"
 
     input:
     tuple val(meta), path(bam)

************************************************************
