Changes in module 'nf-core/qualimap/rnaseq'
--- modules/nf-core/qualimap/rnaseq/main.nf
+++ modules/nf-core/qualimap/rnaseq/main.nf
@@ -2,10 +2,10 @@
     tag "$meta.id"
     label 'process_medium'
 
-    conda "bioconda::qualimap=2.2.2d"
+    conda "bioconda::qualimap=2.3"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/qualimap:2.2.2d--1' :
-        'biocontainers/qualimap:2.2.2d--1' }"
+        'https://depot.galaxyproject.org/singularity/qualimap:2.3--hdfd78af_0' :
+        'biocontainers/qualimap:2.3--hdfd78af_0' }"
 
     input:
     tuple val(meta), path(bam)

************************************************************
